### PR TITLE
Adding popup component (Phase 1)

### PR DESCRIPTION
**Pull request by danielstro**
Initial popup component. Only supports positioning at the moment.
Scrolling / Collision will be dealt in a different PR.

### CODE REVIEW

File: src/components/popup/popup.tsx [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    public render() {
+        if (!this.props.anchor) {
+            return null;
+        }
+        return (
+            <div
+                data-automation-id="POPUP"
+                ref={this.setPosition}
+                style={this.state.style}
+                className={!this.props.open ? style.closed : ''}
```
Comment (tyv) [inline src/components/popup/popup.tsx]: className could be string if you will use Stylable Component

File: src/components/popup/popup.tsx [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    } else {
+        return (window.scrollY ? window.scrollY : window.pageYOffset) + rect[anchorPosition as 'top' | 'bottom'];
+    }
+}
+
+function getHorizontalReference(rect: ClientRect, anchorPosition: HorizontalPosition) {
+    if (anchorPosition === 'center') {
+        return (window.scrollX ? window.scrollX : window.pageXOffset) + rect.left + (rect.width / 2);
+    } else {
+        return (window.scrollX ? window.scrollX : window.pageXOffset) + rect[anchorPosition as 'left' | 'right'];
```
Comment (tyv) [inline src/components/popup/popup.tsx]: you can calculate and save `window.scrollX ? window.scrollX : window.pageXOffset` before `if`

File: src/components/popup/popup.tsx [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            popupStyle.right = `calc(100% - ${getHorizontalReference(anchorRect, anchorPosition)}px`;
+            break;
+    }
+}
+
+function getVerticalReference(rect: ClientRect, anchorPosition: VerticalPosition) {
+    if (anchorPosition === 'center') {
+        return (window.scrollY ? window.scrollY : window.pageYOffset) + rect.top + (rect.height / 2);
+    } else {
+        return (window.scrollY ? window.scrollY : window.pageYOffset) + rect[anchorPosition as 'top' | 'bottom'];
```
Comment (tyv) [inline src/components/popup/popup.tsx]: you can calculate and save `window.scrollY ? window.scrollY : window.pageYOffset` before `if`

File: src/components/popup/popup.tsx [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    syncWidth?: boolean;
+    maxHeight?: number;
+}
+
+export interface PopupState {
+    style: CSSProperties;
+}
+
+export class Popup extends React.Component<Partial<PopupProps>, PopupState> {
+    private static defaultProps: Partial<PopupProps> = {
```
Comment (danielstro) [inline src/components/popup/popup.tsx]: It was an oversight, there is no need for partial here.

File: src/components/popup/popup.tsx (hunk 3 above)
Comment (danielstro) [inline src/components/popup/popup.tsx]: Done.

File: src/components/popup/popup.tsx (hunk 2 above)
Comment (danielstro) [inline src/components/popup/popup.tsx]: Done

File: test/components/popup.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+                expect(select(container, popup)).to.be.absent();
+            });
+            (select(container) as HTMLDivElement).click();
+            return waitForDom(() => {
+                expect(bodySelect(popup)).to.be.present();
+            });
+        });
+    });
+
+    it('renders a hidden pop up', function() {
```
Comment (JoMarton) [inline test/components/popup.spec.tsx]: this test should be removed

File: src/components/popup/popup.tsx [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        super();
+        this.state = { style: {position: 'absolute'}};
+    }
+
+    public render() {
+        if (!this.props.anchor) {
+            return null;
+        }
+        return (
+            <div
```
Comment (JoMarton) [inline src/components/popup/popup.tsx]: Please use the `root` function from wix-react-tools

File: src/components/popup/popup.tsx [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    popupPosition?: PositionPoint;
+    syncWidth?: boolean;
+    maxHeight?: number;
+}
+
+export interface PopupState {
+    style: CSSProperties;
+}
+
+export class Popup extends React.Component<PopupProps, PopupState> {
```
Comment (JoMarton) [inline src/components/popup/popup.tsx]: wrap with `SBComponent(style)`

File: src/components/popup/popup.tsx [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    public render() {
+        if (!this.props.anchor) {
+            return null;
+        }
+        return (
+            <div
+                data-automation-id="POPUP"
+                ref={this.setPosition}
+                style={this.state.style}
+                className={!this.props.open ? style.closed : ''}
```
Comment (JoMarton) [inline src/components/popup/popup.tsx]: use `open` cssState

File: demo/components/popup-demo.tsx [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+import {CSSProperties} from 'react';
+import {Popup} from '../../src/';
+
+export interface DemoState {
+    div: any;
+    isOpen: boolean;
+}
+
+export class PopupDemo extends React.Component<{}, DemoState> {
+    constructor() {
```
Comment (JoMarton) [inline demo/components/popup-demo.tsx]: `state = {...}` will work as well. No need for `constructor()`

File: src/components/popup/popup.tsx [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+        case 'center':
+            popupStyle.left = getHorizontalReference(anchorRect, anchorPosition) - (popupWidth / 2);
+            break;
+        case 'right':
+            popupStyle.left = getHorizontalReference(anchorRect, anchorPosition) - popupWidth;
+            break;
+    }
+}
+
+function getVerticalReference(rect: ClientRect, anchorPosition: VerticalPosition) {
```
Comment (JoMarton) [inline src/components/popup/popup.tsx]: return type

File: src/components/popup/popup.tsx [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+    const yOffset = window.scrollY ? window.scrollY : window.pageYOffset;
+    if (anchorPosition === 'center') {
+        return yOffset + rect.top + (rect.height / 2);
+    } else {
+        return yOffset + rect[anchorPosition as 'top' | 'bottom'];
+    }
+}
+
+function getHorizontalReference(rect: ClientRect, anchorPosition: HorizontalPosition) {
```
Comment (JoMarton) [inline src/components/popup/popup.tsx]: return type

File: src/components/popup/popup.tsx (hunk 10 above)
Comment (danielstro) [inline src/components/popup/popup.tsx]: done

File: src/components/popup/popup.tsx (hunk 11 above)
Comment (danielstro) [inline src/components/popup/popup.tsx]: done

File: test/components/popup.spec.tsx (hunk 5 above)
Comment (danielstro) [inline test/components/popup.spec.tsx]: removed

File: src/components/popup/popup.tsx (hunk 1 above)
Comment (danielstro) [inline src/components/popup/popup.tsx]: I changed the code a bit to use stylable

File: src/components/popup/popup.tsx (hunk 8 above)
Comment (danielstro) [inline src/components/popup/popup.tsx]: Done

File: src/components/popup/popup.tsx (hunk 6 above)
Comment (danielstro) [inline src/components/popup/popup.tsx]: done

File: src/components/popup/popup.tsx (hunk 7 above)
Comment (danielstro) [inline src/components/popup/popup.tsx]: done

File: test/components/popup.spec.tsx [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+import {PopupDemo} from '../../demo/components/popup-demo';
+import {HorizontalPosition, Popup, VerticalPosition} from '../../src/components/';
+
+const popup = 'POPUP';
+const container = 'POPUP_DEMO_DIV';
+
+describe('<Popup />', function() {
+    const clientRenderer = new ClientRenderer();
+    const bodySelect = selectDom(document.body);
+    const testContainer = document.body.appendChild(document.createElement('div'));
```
Comment (AviVahl) [inline test/components/popup.spec.tsx]: if this is your test anchor, name it as such.
why is this here and not inside the test?

File: test/components/popup.spec.tsx [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+            await waitForDom(() => {
+                expect(bodySelect(popup)).to.be.present();
+            });
+            (select(container) as HTMLDivElement).click();
+            return waitForDom(() => {expect(bodySelect(popup)).to.be.absent(); });
+        });
+    });
+
+    it('displays the popup and renders its children if the open prop is given', function() {
+        const div = document.body.appendChild(testContainer);
```
Comment (AviVahl) [inline test/components/popup.spec.tsx]: you are re-adding the anchor?

File: test/components/popup.spec.tsx [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+
+    describe('The popup user', function() {
+        it('clicks on the parent and the popup opens and closes after another click', async function() {
+            const {select, waitForDom} = clientRenderer.render(<PopupDemo />);
+
+            await waitForDom(() => {
+                expect(select(container)).to.be.present();
+                expect(select(container, popup)).to.be.absent();
+            });
+            (select(container) as HTMLDivElement).click();
```
Comment (AviVahl) [inline test/components/popup.spec.tsx]: select is a generic function. it accepts the target's type (extends `Element`). Use it instead of type casting.

File: test/components/popup.spec.tsx [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+        clientRenderer.render(
+            <Popup
+                anchor={div!}
+                open={true}
+            >
+                <span data-automation-id="SPAN">Popup Body</span>
+            </Popup>);
+
+        await waitFor(() => {expect(bodySelect(popup)).to.be.present(); });
+        ReactDOM.unmountComponentAtNode(bodySelect(popup)!.parentElement!);
```
Comment (AviVahl) [inline test/components/popup.spec.tsx]: this is not how we do it.
You should create a rendering container in the test. Pass it to clientRenderer.render (second parameter), and only then you are allowed to unmount it yourself.
You should verify the current code does not crash outside the test (unmounting an already unmounted component).

File: test/components/popup.spec.tsx [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+        const horizontalArray = ['left', 'center', 'right'];
+
+        // Level one: popup position, level two: anchor position
+        const topResults = getLayoutTest('vertical');
+        const leftResults = getLayoutTest('horizontal');
+
+        verticalArray.forEach((popupVertical: VerticalPosition) => {
+            horizontalArray.forEach((popupHorizontal: HorizontalPosition) => {
+                verticalArray.forEach((anchorVertical: VerticalPosition) => {
+                    horizontalArray.forEach((anchorHorizontal: HorizontalPosition) => {
```
Comment (AviVahl) [inline test/components/popup.spec.tsx]: wtf?

File: src/components/popup/popup.tsx [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+    public componentDidMount() {
+        this.container = document.body.appendChild(document.createElement('div'));
+        if (this.props.anchor) {
+            ReactDOM.unstable_renderSubtreeIntoContainer(this, this.popup, this.container);
+        }
+    }
+
+    public componentDidUpdate() {
+        if (this.props.anchor && this.container && this.popup) {
+            const x = ReactDOM.unstable_renderSubtreeIntoContainer(this, this.popup, this.container);
```
Comment (danielstro) [inline src/components/popup/popup.tsx]: My bad, I added to try and debug the IE11 problem and I didn't remove it.

File: src/components/popup/popup.tsx [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+            break;
+        case 'right':
+            popupStyle.left = getHorizontalReference(anchorRect, anchorPosition);
+            addTransform(popupStyle, 'translateX(-100%)');
+            break;
+    }
+}
+
+function getVerticalReference(rect: ClientRect, anchorPosition: VerticalPosition): number {
+    const yOffset = window.scrollY ? window.scrollY : window.pageYOffset;
```
Comment (danielstro) [inline src/components/popup/popup.tsx]: IE & Safari don't support scrollY (or partially support), I will change the statement to check if it's undefined.

File: src/components/popup/popup.st.css [hunk 19]
```diff
@@ -0,0 +1,9 @@
+@namespace "Popup";
+
+.root {
+    -st-states: open;
+}
+.root:not(:open) {
```
Comment (danielstro) [inline src/components/popup/popup.st.css]: It's not, I forgot to remove this file.

File: src/components/popup/popup.tsx [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+        setLeft(newStyle, anchorRect,
+            anchorHorizontalPos, this.props.popupPosition!.horizontal);
+        if (this.props.syncWidth) {
+            newStyle.width = anchorRect.width;
+        }
+        return newStyle;
+    }
+}
+
+function setTop(popupStyle: React.CSSProperties, anchorRect: ClientRect,
```
Comment (tobich) [inline src/components/popup/popup.tsx]: I suggest rewriting to even more functional style. IMHO it can be simplified more.

File: src/components/popup/popup.tsx [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+        case 'bottom':
+            popupStyle.top = getVerticalReference(anchorRect, anchorPosition);
+            addTransform(popupStyle, 'translateY(-100%)');
+            break;
+    }
+}
+
+function setLeft(popupStyle: React.CSSProperties, anchorRect: ClientRect,
+                 anchorPosition: HorizontalPosition, popupPoistion: HorizontalPosition) {
+    switch (popupPoistion) {
```
Comment (tobich) [inline src/components/popup/popup.tsx]: typo

File: src/components/popup/popup.tsx [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+                <Portal style={this.createStyle()}>
+                    {this.props.children}
+                </Portal>);
+        }
+
+        return null;
+    }
+
+    private createStyle() {
+        if (!this.props.anchor) {
```
Comment (tobich) [inline src/components/popup/popup.tsx]: This condition is superfluous

File: src/components/popup/popup.tsx [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+    public static defaultProps: Partial<PopupProps> = {
+        open: false,
+        anchorPosition: {vertical: 'bottom', horizontal: 'left'},
+        popupPosition: {vertical: 'top', horizontal: 'left'},
+        syncWidth: true,
+        maxHeight: 500
+    };
+
+    public render() {
+        if (this.props.anchor && this.props.open) {
```
Comment (tobich) [inline src/components/popup/popup.tsx]: Please add test and `| null` typing for the `this.props.anchor` as null (valid use-case, should be tested).

File: src/components/popup/popup.tsx [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+            return (
+                <Portal style={this.createStyle()}>
+                    {this.props.children}
+                </Portal>);
+        }
+
+        return null;
+    }
+
+    private createStyle() {
```
Comment (tobich) [inline src/components/popup/popup.tsx]: Please add return type.
